File: java/BinarySearch.java
// time complexity; O(log n)
// space complexity; O(1)

public class BinarySearch
  {
    public static int binarySearch(int[] arr, int x)
    {
      int low = 0, high = arr.length - 1;
      while (low <= high)
        {
          int mid = low + (high - low) / 2;
          // check if x is present at mid
          if (arr[mid] == x)
            return mid;
          // if x greater, ignore left half
          if (arr[mid] < x)
            low = mid + 1;
          // if x smaller, ignore right half
          else
            high = mid - 1;
        }
      // if we reach here, element not present
    return -1;
    }
  }

File: python/binary_search.py
# time complexity; O(log n)
# space complexity; O(1)

def binary_search(arr, x):
  low = 0
  high = len(arr) - 1
  mid = 0

  while low <= high:
    mid = (high + low) // 2

    #if x is greater, ignore left half
    if arr[mid] < x:
      low = mid + 1
    # if x is smaller, ignore right half
    elif arr[mid] > x:
      high = mid - 1
    else: #otherwise, element is present at mid 
      return mid
  return -1 # if we reach here, element wasn't present 
